File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Helpers/LoadableSource.kt
package com.abracode.actionui.Helpers

import kotlinx.serialization.json.JsonObject

/**
 * The serialized format of a loaded description, inferred from the source's file
 * extension. `.plist` selects [PLIST] (Apple binary/XML property list); anything
 * else - including no extension - is [JSON].
 *
 * On Android only [JSON] is decodable today: there is no built-in property-list
 * parser, so a [PLIST] source warn-and-skips (the same honest degrade pattern the
 * rest of the port uses). The enum is carried through so the deferred plist path
 * is explicit rather than silently misread as JSON.
 */
internal enum class LoadableFormat { JSON, PLIST }

/**
 * A `LoadableView` source string classified by *how* it must be fetched. Pure data
 * produced by [classifyLoadableSource]; the renderer ([com.abracode.actionui.Views.LoadableView])
 * turns each case into the actual read.
 *
 * The heuristics mirror Apple's `LoadableView.buildView`
 * (`ActionUI/Views/LoadableView.swift`): an `http://` / `https://` prefix is a
 * remote URL, a `file://` prefix or any string containing `/` is a filesystem
 * path, and anything else is a bundle resource name.
 */
internal sealed interface LoadableSource {
    /** No usable source (null / blank). Renders the "no source" placeholder. */
    object None : LoadableSource

    /** Bundle resource by [name] - read from the app's `assets/` on Android. */
    data class BundleName(val name: String, val format: LoadableFormat) : LoadableSource

    /** Absolute filesystem [path] (any `file://` scheme already stripped). */
    data class FilePath(val path: String, val format: LoadableFormat) : LoadableSource

    /**
     * Remote `http(s)` [url]. Deferred on Android (the synchronous local-source
     * chunk ships first; remote async loading lands with `RemoteLoadableView`),
     * so the renderer warn-and-skips this case.
     */
    data class Remote(val url: String, val format: LoadableFormat) : LoadableSource
}

/**
 * The effective source string for a `LoadableView`: the runtime [value] when set
 * and non-blank, else the first present and non-blank of `url` > `filePath` >
 * `name` from [properties]. Null when none is available. Mirrors the priority in
 * Apple's `LoadableView.initialValue`. Pure.
 */
internal fun resolveLoadableSource(value: String?, properties: JsonObject?): String? {
    if (!value.isNullOrBlank()) return value
    for (key in arrayOf("url", "filePath", "name")) {
        val candidate = properties?.stringProperty(key)
        if (!candidate.isNullOrBlank()) return candidate
    }
    return null
}

/**
 * Classifies [source] into a [LoadableSource] by Apple's heuristics (see that
 * type). A blank / null source is [LoadableSource.None]. The [LoadableFormat] is
 * read from the trailing extension (`.plist` -> [LoadableFormat.PLIST], else
 * [LoadableFormat.JSON]). Pure and unit-tested.
 */
internal fun classifyLoadableSource(source: String?): LoadableSource {
    val trimmed = source?.trim().orEmpty()
    if (trimmed.isEmpty()) return LoadableSource.None
    val lower = trimmed.lowercase()
    return when {
        lower.startsWith("http://") || lower.startsWith("https://") ->
            LoadableSource.Remote(trimmed, formatOf(trimmed))
        lower.startsWith("file://") -> {
            val path = trimmed.substring("file://".length)
            LoadableSource.FilePath(path, formatOf(path))
        }
        trimmed.contains('/') -> LoadableSource.FilePath(trimmed, formatOf(trimmed))
        else -> LoadableSource.BundleName(trimmed, formatOf(trimmed))
    }
}

/** [LoadableFormat] from a source's trailing extension; defaults to JSON. Pure. */
private fun formatOf(source: String): LoadableFormat =
    if (source.substringAfterLast('.', "").lowercase() == "plist") LoadableFormat.PLIST
    else LoadableFormat.JSON

File: ActionUIWeb/src/Helpers/LoadableSource.js
// LoadableSource.js — classify a LoadableView source string.
// Web analog of ActionUIAndroid Helpers/LoadableSource.kt (the pure
// resolve/classify functions), adapted to produce a browser fetch URL.
//
// A LoadableView's source can arrive three ways (Apple's `url` > `filePath` >
// `name` priority); the string is then classified by the same heuristics on every
// platform — an http(s) prefix is a remote URL, a `file://` prefix or any string
// containing `/` is a path, and anything else is a bundle/resource name. The
// serialized format is read from the trailing extension (`.plist` → "plist", else
// "json").
//
// On the web there is no app bundle or synchronous filesystem: every kind resolves
// to a `fetch(url)` relative to the host page. So classify carries a `fetchURL` —
// the remote URL as-is, the path (with any `file://` stripped), or the bundle name
// with `.json` appended when it has no extension — alongside the `kind` (kept for
// parity/diagnostics) and `format`.

// The effective source string: the runtime [value] when set and non-blank, else
// the first present and non-blank of url > filePath > name. Null when none.
export function resolveLoadableSource(value, properties) {
    if (typeof value === "string" && value.trim() !== "") return value;
    for (const key of ["url", "filePath", "name"]) {
        const candidate = properties?.[key];
        if (typeof candidate === "string" && candidate.trim() !== "") return candidate;
    }
    return null;
}

// Classifies a source into { kind, fetchURL, format }. A blank/absent source is
// { kind: "none" }. kind ∈ "none"|"remote"|"filePath"|"bundle".
export function classifyLoadableSource(source) {
    const trimmed = typeof source === "string" ? source.trim() : "";
    if (trimmed === "") return { kind: "none", fetchURL: "", format: "json" };

    const lower = trimmed.toLowerCase();
    if (lower.startsWith("http://") || lower.startsWith("https://")) {
        return { kind: "remote", fetchURL: trimmed, format: formatOf(trimmed) };
    }
    if (lower.startsWith("file://")) {
        const path = trimmed.slice("file://".length);
        return { kind: "filePath", fetchURL: path, format: formatOf(path) };
    }
    if (trimmed.includes("/")) {
        return { kind: "filePath", fetchURL: trimmed, format: formatOf(trimmed) };
    }
    // Bundle/resource name — on the web, fetched relative to the page; append
    // `.json` when the name carries no extension (Android appends it too).
    const fetchURL = trimmed.includes(".") ? trimmed : `${trimmed}.json`;
    return { kind: "bundle", fetchURL, format: formatOf(fetchURL) };
}

// The serialized format from a source's trailing extension; defaults to "json".
function formatOf(source) {
    const ext = source.includes(".") ? source.slice(source.lastIndexOf(".") + 1).toLowerCase() : "";
    return ext === "plist" ? "plist" : "json";
}
